suite/spdz2k/src/main/java/dk/alexandra/fresco/suite/marlin/protocols/computations/Spdz2kMacCheckComputation.java

```diff
@@ -0,0 +1,106 @@
+    PlainT extends CompUInt<HighT, LowT, PlainT>>
+    implements Computation<Void, ProtocolBuilderNumeric> {
+
+  private final Spdz2kResourcePool<PlainT> resourcePool;
+  private final CompUIntConverter<HighT, LowT, PlainT> converter;
+
+  public Spdz2kMacCheckComputation(Spdz2kResourcePool<PlainT> resourcePool,
+      CompUIntConverter<HighT, LowT, PlainT> converter) {
+    this.resourcePool = resourcePool;
+    this.converter = converter;
+  }
+
+  @Override
+  public DRes<Void> buildComputation(ProtocolBuilderNumeric builder) {
+    Pair<List<Spdz2kSInt<PlainT>>, List<PlainT>> opened = resourcePool
+        .getOpenedValueStore()
+        .popValues();
+    List<Spdz2kSInt<PlainT>> authenticatedElements = opened.getFirst();
+    List<PlainT> openValues = opened.getSecond();
+    ByteSerializer<PlainT> serializer = resourcePool.getRawSerializer();
+    CompUIntFactory<PlainT> factory = resourcePool.getFactory();
+    PlainT macKeyShare = resourcePool.getDataSupplier().getSecretSharedKey();
+    List<byte[]> sharesLowBits = authenticatedElements.stream()
+        .map(element -> element.getShare().getLeastSignificant().toByteArray())
+        .collect(Collectors.toList());
+    final List<PlainT> randomCoefficients = sampleCoefficients(
+        resourcePool.getRandomGenerator(),
+        factory, openValues.size());
+    final PlainT y = UInt.innerProduct(openValues, randomCoefficients);
+    final Spdz2kSInt<PlainT> r = resourcePool.getDataSupplier()
+        .getNextRandomElementShare();
+    return builder
+        .seq(new BroadcastComputation<>(sharesLowBits))
+        .seq((seq, ignored) -> {
+          List<PlainT> originalShares = authenticatedElements.stream()
+              .map(Spdz2kSInt::getShare)
+              .collect(Collectors.toList());
+          List<HighT> overflow = originalShares.stream()
+              .map(PlainT::computeOverflow)
+              .collect(Collectors.toList());
+          List<HighT> randomCoefficientsAsHigh = randomCoefficients.stream()
+              .map(PlainT::getLeastSignificantAsHigh)
+              .collect(Collectors.toList());
+          HighT pj = UInt.innerProduct(overflow, randomCoefficientsAsHigh);
+          byte[] pjBytes = pj.add(r.getShare().getLeastSignificantAsHigh()).toByteArray();
+          return new BroadcastComputation<ProtocolBuilderNumeric>(pjBytes).buildComputation(seq);
+        })
+        .seq((seq, broadcastPjs) -> {
+          List<PlainT> pjList = serializer.deserializeList(broadcastPjs);
+          HighT pLow = UInt.sum(
+              pjList.stream().map(PlainT::getLeastSignificantAsHigh).collect(Collectors.toList()));
+          PlainT p = converter.createFromHigh(pLow);
+          List<PlainT> macShares = authenticatedElements.stream()
+              .map(Spdz2kSInt::getMacShare)
+              .collect(Collectors.toList());
+          PlainT mj = UInt.innerProduct(macShares, randomCoefficients);
+          PlainT zj = macKeyShare.multiply(y)
```

Comment:
Strange varnames should either reference to a paper or documentation outside of the code. OR be renamed